src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/DefaultWindowsRazorProjectHost.cs

```diff
@@ -58,93 +53,63 @@ internal class DefaultWindowsRazorProjectHost : WindowsRazorProjectHostBase
     {
     }
 
-    protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
-    {
-        await base.InitializeCoreAsync(cancellationToken).ConfigureAwait(false);
-
-        // Don't try to evaluate any properties here since the project is still loading and we require access
-        // to the UI thread to push our updates.
-        //
-        // Just subscribe and handle the notification later.
-        var receiver = new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(OnProjectChangedAsync);
-        _subscription = CommonServices.ActiveConfiguredProjectSubscription.JointRuleSource.SourceBlock.LinkTo(
-            receiver,
-            initialDataAsNew: true,
-            suppressVersionOnlyUpdates: true,
-            ruleNames: new string[]
-            {
-                Rules.RazorGeneral.SchemaName,
-                Rules.RazorConfiguration.SchemaName,
-                Rules.RazorExtension.SchemaName,
-                Rules.RazorComponentWithTargetPath.SchemaName,
-                Rules.RazorGenerateWithTargetPath.SchemaName,
-                ConfigurationGeneralSchemaName,
-            });
-    }
-
-    protected override async Task DisposeCoreAsync(bool initialized)
+    protected override string[] GetRuleNames()
```

Comment:
nit: do we need to create a new array every time?

Reply:
As commented below (and after your comment here) CPS internally uses `ImmutableHashSet<string>` for these values. You can avoid having CPS create one of those "expensive" structures each time (and get better immutability guarantees here too) by making your static field here an `ImmutableHashSet<string>` rather than a `string[]`.